workspaces/ui-v2/src/store/endpoints.ts

```diff
@@ -0,0 +1,127 @@
+import {
+  createSlice,
+  createAsyncThunk,
+  SerializedError,
+} from '@reduxjs/toolkit';
+import { IForkableSpectacle } from '@useoptic/spectacle';
+import { AsyncStatus, IEndpoint } from '<src>/types';
+
+const initialState: {
+  data: AsyncStatus<IEndpoint[], SerializedError>;
+} = {
+  data: {
+    loading: true,
+  },
+};
+
+function sharedStart(array: string[]): string {
+  if (array.length === 0) return '/';
+  let A = array.concat().sort(),
+    a1 = A[0],
+    a2 = A[A.length - 1],
+    L = a1.length,
+    i = 0;
+  while (i < L && a1.charAt(i) === a2.charAt(i)) i++;
+  return a1.substring(0, i);
+}
+
+const endpointsSlice = createSlice({
+  name: 'endpoints',
+  initialState,
+  reducers: {},
+  extraReducers: (builder) => {
+    builder.addCase(fetchEndpoints.fulfilled, (state, action) => {
+      const results = action.payload;
+
+      state.data = {
+        loading: false,
+        data: endpointQueryResultsToJson(results),
+      };
+    });
+    builder.addCase(fetchEndpoints.rejected, (state, action) => {
+      state.data = {
+        loading: false,
+        error: action.error,
+      };
+    });
+  },
+});
+
+export const AllEndpointsQuery = `{
+  requests {
+    id
+    pathId
+    absolutePathPattern
+    absolutePathPatternWithParameterNames
+    pathComponents {
+      id
+      name
+      isParameterized
+      contributions
+    }
+    method
+    pathContributions
+  }
+}`;
+
+export type EndpointQueryResults = {
+  requests: {
+    id: string;
+    pathId: string;
+    absolutePathPattern: string;
+    absolutePathPatternWithParameterNames: string;
+    pathComponents: {
+      id: string;
+      name: string;
+      isParameterized: boolean;
+      contributions: Record<string, string>;
+    }[];
+    method: string;
+    pathContributions: Record<string, string>;
+  }[];
+};
+
+export const endpointQueryResultsToJson = ({
+  requests,
+}: EndpointQueryResults) => {
+  const commonStart = sharedStart(
+    requests.map((req) => req.absolutePathPatternWithParameterNames)
+  );
+
+  return requests.map((request) => ({
+    pathId: request.pathId,
+    method: request.method,
+    fullPath: request.absolutePathPatternWithParameterNames,
+    group: request.absolutePathPatternWithParameterNames
+      .substring(commonStart.length)
+      .split('/')[0],
+    pathParameters: request.pathComponents.map((path) => ({
+      id: path.id,
+      name: path.name,
+      isParameterized: path.isParameterized,
+      description: path.contributions.description || '',
+      endpointId: `${request.pathId}.${request.method}`,
+    })),
+    description: request.pathContributions.description || '',
+    purpose: request.pathContributions.purpose || '',
+  }));
+};
+
+const fetchEndpoints = createAsyncThunk(
+  'FETCH_ENDPOINTS',
+  async ({ spectacle }: { spectacle: IForkableSpectacle }) => {
```

Comment:
I think this makes the most sense to do dependency injection here rather than trying to include this in middleware

Reply:
Especially true if the instance can change! I've made the mistake of writing things as middleware more than once before, and have regretted it. There's definite reasons to going the middleware route, but I've learnt the hard way that defaulting to simple dependency injection is most likely the best path.